### PR TITLE
markdown: fix link and color interpolation, refactor error reporting

### LINKED ISSUE #11527: StyledText don't render dynamic LINKS with @markdown()
### Bug Description

There is no clear supported syntax for using dynamic values inside `@markdown(...)` in `StyledText`. I expected to be able to build a markdown link from a property or global value, but the current behavior is confusing and undocumented.

**Environment**
- Slint UI framework
- `StyledText` / `@markdown`
- `LibrariesData` global string values

This issue is about making dynamic markdown usage more intuitive and preventing false assumptions in UI code.

### Reproducible Code (if applicable)

```slint
**Reproduction**


StyledText {
    text: @markdown("[Link](" + LibrariesData.library_url + ")");
}


or


StyledText {
    text: @markdown("[Link]({LibrariesData.library_url})");
}


or


StyledText {
    text: @markdown("[Link](\{LibrariesData.library_url})");
}
```

### Environment Details

- Slint Version: 1.16.1
- Platform/OS: WIndows 10 22H2
- Programming Language: Rust + Slint
- Backend/Renderer:

### Product Impact

- I am rewriting the [`Bridgex`](https://github.com/dev2forge/bridgex) project in Slint, which is the GUI for file-to-Markdown conversion at `https://dev2forge.is-a.software/bridgex/`.
- I am building a component that must render dynamic links, driven by data such as `LibrariesData.library_url`.
- This issue is a genuine inconvenience because it blocks correct implementation of the dynamic links component and affects the user experience directly.

### PR DESCRIPTION
**Pull request by ogoffart**
The first commit fixes #11527: interpolated values in markdown link URLs were not substituted. (This one can be backported to the release branch if we need to)
  
The second commit adds support for color interpolation (`<font color="\{expr}">`). This requires parse_interpolated to always produce output instead of stopping at the first error, so invalid colors at runtime still render text and log the error. The refactor also adds byte ranges to errors for precise source locations in diagnostics, suppresses cascading errors, and cleans up error messages. 

(The third commit is just a test for the debug_log output)

### CODE REVIEW

**Comment by ogoffart** [comment]
(CI error unrelated)

File: internal/common/styled_text.rs [hunk 1]
```diff
@@ -138,58 +138,88 @@ pub fn parse_interpolated<S: AsRef<[StyledTextParagraph]>>(
...
-                                },
-                            ));
-                        }
-                        [] => {
-                            // No paragraphs in the argument, so nothing to add
-                        }
-                        _ => {
-                            return Err(StyledTextError::MultiParagraphInterpolation);
-                        }
+    fn substitute<S: AsRef<[StyledTextParagraph]>>(
```
Comment (LeonMatthes) [inline internal/common/styled_text.rs]: I'm all for breaking this up into descriptive functions, but I think it's very confusing that the functions are all within the parse function. Please move them out and place them in the module. Otherwise the function is extremely hard to read - with some code at the start, then multiple large functions and then the remaining code.

File: internal/common/styled_text.rs [hunk 2]
```diff
@@ -42,11 +42,50 @@ pub struct StyledTextParagraph {
     pub links: alloc::vec::Vec<(core::ops::Range<usize>, alloc::string::String)>,
 }
 
-/// Error type returned by `StyledText::parse`
+/// Error returned by markdown styled text parsing
 #[cfg(feature = "markdown")]
-#[derive(Debug, derive_more::Error, derive_more::Display, PartialEq)]
-#[non_exhaustive]
-pub enum StyledTextError<'a> {
```
Comment (LeonMatthes) [inline internal/common/styled_text.rs:49]: Just to be sure, this type was not part of the public API before, right?
If I recall correctly, creating style text from Rust or any other language was not released with 1.16.

Otherwise renaming it would be a breaking change.

File: internal/common/styled_text.rs [hunk 3]
```diff
@@ -42,11 +42,50 @@ pub struct StyledTextParagraph {
...
+    }
+
+    fn without_range(kind: StyledTextParseErrorKind) -> Self {
+        Self { kind, range: None }
+    }
+
+    /// Returns true if this is an invalid color error.
+    /// Useful for the compiler to skip this during compile-time validation
+    /// when color values may come from dynamic interpolation.
+    pub fn is_invalid_color(&self) -> bool {
```
Comment (LeonMatthes) [inline internal/common/styled_text.rs]: I can see how this is useful to the compiler, but should this be public API on this type? Because that will be returned by the API that end users are going to use as well. Right?

Maybe it's better to make this stand-alone function that the compiler can access but is not part of the Slint crate.

File: internal/common/styled_text.rs [hunk 4]
```diff
@@ -206,312 +250,413 @@ pub fn parse_interpolated<S: AsRef<[StyledTextParagraph]>>(
...
-            match event {
-                pulldown_cmark::Event::SoftBreak | pulldown_cmark::Event::HardBreak => {
-                    if let Some(paragraph) =
-                        current_paragraph.replace(begin_paragraph(indentation, None))
-                    {
-                        return Some(Ok(paragraph));
-                    }
+    fn unsupported_tag_name(tag: &pulldown_cmark::Tag<'_>) -> alloc::string::String {
+        match tag {
+            pulldown_cmark::Tag::Heading { .. } => "headings",
```
Comment (LeonMatthes) [inline internal/common/styled_text.rs]: nit: These could really use a `use pulldown_cmark::Tag::*`.

File: internal/common/styled_text.rs [hunk 5]
```diff
@@ -206,312 +250,413 @@ pub fn parse_interpolated<S: AsRef<[StyledTextParagraph]>>(
...
+            pulldown_cmark::Tag::Superscript => "superscript",
+            pulldown_cmark::Tag::Subscript => "subscript",
+            other => return alloc::format!("{:?}", other.to_end()),
+        }
+        .into()
+    }
+
+    fn unsupported_event_name(event: &pulldown_cmark::Event<'_>) -> alloc::string::String {
+        match event {
+            pulldown_cmark::Event::Rule => "horizontal rules".into(),
```
Comment (LeonMatthes) [inline internal/common/styled_text.rs]: nit: `use pulldown_cmark::Event::*;`

File: internal/compiler/passes/resolving.rs [hunk 6]
```diff
@@ -813,13 +844,50 @@ impl Expression {
...
+            // Look up the source location from the error's byte range.
+            // Compute sub-literal precision: adjust the source span to point
+            // at the specific position within the string literal.
+            let loc = e.range().and_then(|r| {
+                let idx = source_map.partition_point(|(range, _)| range.start <= r.start);
+                if idx > 0 {
+                    let (fmt_range, loc) = &source_map[idx - 1];
+                    if fmt_range.contains(&r.start) {
+                        let delta = r.start - fmt_range.start;
+                        let err_len = (r.end - r.start).min(loc.span.length.saturating_sub(delta));
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs]: Use `range.len()` from its ExactSizeIterator impl instead of `r.end - r.start`

File: internal/compiler/passes/resolving.rs [hunk 7]
```diff
@@ -813,13 +844,50 @@ impl Expression {
...
+            {
+                continue;
+            }
+            // Look up the source location from the error's byte range.
+            // Compute sub-literal precision: adjust the source span to point
+            // at the specific position within the string literal.
+            let loc = e.range().and_then(|r| {
+                let idx = source_map.partition_point(|(range, _)| range.start <= r.start);
+                if idx > 0 {
+                    let (fmt_range, loc) = &source_map[idx - 1];
```
Comment (LeonMatthes) [inline internal/compiler/passes/resolving.rs:872]: Why the `-1` here? Shouldn't the partition point function give us the index of the first range that contains r.start?

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 8]
```diff
@@ -5,35 +5,112 @@ export component X {
     property <styled-text> t1: @markdown("Foo\invalid");
 //                                       >           <error{Cannot parse string literal}
     property <styled-text> t2: @markdown("Foo\u{e541}Bar");  // Using the MARKDOWN_INTERPOLATION_PLACEHOLDER cause strange errors
-//                             >                         <error{Argument index 0 out of range: 0 arguments provided}
+//                                        >       <error{Argument index 0 out of range: 0 arguments provided}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint:8]: This produces two errors for the same issue. The range of the error is not correct and the argument index error is very confusing.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 9]
```diff
@@ -5,35 +5,112 @@ export component X {
...
+//                             >                         <^error{Format string contains 1 placeholders, but 0 arguments were provided}
     property <styled-text> t3: @markdown("XXX"  "TTT \{12px} ZZZ");
 //                                                     >  <error{Cannot convert length to string. Divide by 1px to convert to a plain number}
     property <styled-text> t9: @markdown("{");
     property <styled-text> t10: @markdown("}");
     property <styled-text> t11: @markdown("{abc}");
     property <styled-text> t12: @markdown("{0");
     property <styled-text> t13: @markdown("> quote");
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint:16]: To test the error handling more thoroughly,it would be interesting to combine more of these errors into a single `@markdown` invocation.
Then we would know that the offsetting works correctly even for longer text segments and across multiple paragraphs.

And we would reduce a lot of the boilerplate in terms of having a bunch of properties individually.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 10]
```diff
@@ -5,35 +5,112 @@ export component X {
...
     property <styled-text> t15: @markdown("---");
-//                              >              <error{Unimplemented event: Rule}
+//                                         > <error{Markdown horizontal rules are not supported}
     property <styled-text> t16: @markdown("<span>text</span>");
-//                              >                            <error{Unimplemented html tag: span}
+//                                         >   <error{HTML tag <span> is not supported}
     property <styled-text> t17: @markdown("<div>text</div>");
-//                              >                          <error{Unimplemented tag: HtmlBlock}
+//                                         >             <error{Markdown HTML blocks are not supported}
+//                                         >             <^error{Markdown HTML blocks (<div>text</div>) are not supported}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint]: Unnecessary duplicate error

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 11]
```diff
@@ -5,35 +5,112 @@ export component X {
...
     property <styled-text> t20: @markdown("<u>text</font>");
-//                              >                         <error{Closing html tag doesn't match the opening tag. Expected </u>, got </font>}
+//                                                >     <error{Closing html tag doesn't match the opening tag. Expected </u>, got </font>}
     property <styled-text> t21: @markdown("<font>text</font>");
-//                              >                            <error{Missing color attribute in html <font>}
+//                                         >    <error{Missing color attribute in html <font>}
     property <styled-text> t22: @markdown("<font size='12'>text</font>");
-//                              >                                      <error{Unexpected size attribute in html <font size='12'>}
+//                                               >       <error{Unexpected size attribute in html <font size='12'>}
+//                                         >              <^error{Missing color attribute in html <font size='12'>}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint]: This error is also duplicate and confusing.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 12]
```diff
@@ -5,35 +5,112 @@ export component X {
...
+//                                              >  <error{HTML tag <xyz> is not supported}
+
+    // Valid markdown mixed with errors in different literals
+    property <styled-text> m5: @markdown("**bold** " "<xyz>bad</xyz>" " *italic*");
+//                                                    >  <error{HTML tag <xyz> is not supported}
+
+    // Block quote error in a multi-line expression
+    property <styled-text> m6: @markdown("text"
+                                         "\n> block quote");
+//                                         >           <error{Markdown block quotes are not supported}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint:69]: This seems to be the wrong location, shouldn't it start at the `>` instead of the `\n`?

File: internal/core/styled_text.rs
Comment (LeonMatthes) [inline internal/core/styled_text.rs:83]: Instead of silently failing, maybe this whole function should be unavailable if std is not available?

**Review by LeonMatthes — COMMENTED** [review]
I have to admit that it is hard to tell if any of the refactoring may have broken something, but in general this is going in a good direction.

Some of the ranges on the errors don't seem quite right and some are duplicated/confusing, so those should be adjusted.

File: internal/common/styled_text.rs (hunk 2 above)
Comment (ogoffart) [reply internal/common/styled_text.rs:49]: Yes, this is not yet public API.
This is actually going to be made public with https://github.com/slint-ui/slint/pull/11180 which conflict with this PR though.

File: internal/common/styled_text.rs (hunk 3 above)
Comment (ogoffart) [reply internal/common/styled_text.rs]: Indeed, i've done it this way now because it was not public API, but as the type will become public in the future, we should be prepared.

An extra function is a good solution.

(I was actually thinking of some other followup regarding color which we can talk about later)

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint (hunk 8 above)
Comment (ogoffart) [reply internal/compiler/tests/syntax/basic/markdown_syntax.slint:8]: Well, this error is a corner case, nobody is going to see that.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint (hunk 12 above)
Comment (ogoffart) [reply internal/compiler/tests/syntax/basic/markdown_syntax.slint:69]: The problem here is that escape sequences within string break the location map.
The fix is going to be complicated as we'd need to account for it in the mapping.

File: internal/core/styled_text.rs
Comment (ogoffart) [reply internal/core/styled_text.rs:83]: Indeed.  This is pre-existing.  I wonder if this is a breaking change or not.

File: internal/common/styled_text.rs (hunk 1 above)
Comment (ogoffart) [reply internal/common/styled_text.rs]: I like to put function that are helper to a single function within its scope. But it is true that I don't do that consistently.

File: internal/compiler/passes/resolving.rs (hunk 7 above)
Comment (ogoffart) [reply internal/compiler/passes/resolving.rs:872]: partition_point returns the first index where range.start > r.start, so idx - 1 is the last entry whose range could contain r.start.

(idx = 0 would mean that the error is before the first range, which is presumably impossible)

**Review by ogoffart — COMMENTED** [review]
Thanks for the review. I addressed most comments

I think the following could be done in a follow-up:
 - fix error location if there are escape sequences 
 - remove the non-working functions in no_std

File: internal/common/styled_text.rs (hunk 1 above)
Comment (LeonMatthes) [reply internal/common/styled_text.rs]: Usually I do this as well, but that only works up to a given size of functions. So for anything larger, the solution is usually to split it into multiple separate functions and move those into a module, which is basically what we have here already. :)

File: internal/compiler/passes/resolving.rs (hunk 7 above)
Comment (LeonMatthes) [reply internal/compiler/passes/resolving.rs:872]: Ah true. This is a bit confusing as it's basically the opposite of what the condition says, but it makes sense. Thanks for adding the comment  :+1:

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint (hunk 12 above)
Comment (LeonMatthes) [reply internal/compiler/tests/syntax/basic/markdown_syntax.slint:69]: Hm, fair enough...
(Please don't resolve comments that you have answered, but not fixed in the code, as I want to read and potentially reply to your reasoning 🙂 )

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint (hunk 8 above)
Comment (LeonMatthes) [reply internal/compiler/tests/syntax/basic/markdown_syntax.slint:8]: ah, right, that only happens if you manually add the special unicode character. fair enough.

File: internal/core/styled_text.rs
Comment (LeonMatthes) [reply internal/core/styled_text.rs:83]: I thought we didn't yet expose markdown parsing in the public API so it shouldn't be a breaking change. If it is not, please address this now before it slips into public API.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint [hunk 13]
```diff
@@ -113,4 +111,10 @@ export component X {
                                           "<abc>bad</abc>"
 //                                         >  <error{HTML tag <abc> is not supported}
                                           " suffix");
+
+    // Combined errors from different markdown features in a single string with escape sequences
+    property <styled-text> t25: @markdown("---\n\n> quote\n\n<span>text</span>");
+//                                         >  <error{Markdown horizontal rules are not supported}
+//                                              >      <^error{Markdown block quotes are not supported}
```
Comment (LeonMatthes) [inline internal/compiler/tests/syntax/basic/markdown_syntax.slint:118]: The error about the quotes is again not spanning the right part of the string. You mentioned before that that is due to the replacement character, but I don't think there is a replacement character involved here, right?

**Review by LeonMatthes — APPROVED** [review]
Still think the spans on the quotes are not quite right, even if no placeholder is involved. Other than that, I think this is okay to merge.

File: internal/compiler/tests/syntax/basic/markdown_syntax.slint (hunk 13 above)
Comment (ogoffart) [reply internal/compiler/tests/syntax/basic/markdown_syntax.slint:118]: Yes, it is because of the `\n`  will fix up in another commit.
